### PR TITLE
Pipes -- Diane Lenertz -- APIMuncher

### PR DESCRIPTION
**Pull request by lenertzdiane**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? |  I read the documentation in order to find how to query the API correctly|
| Describe your API Wrapper.  How did you decide on the methods you created? | I knew I needed one to list all the recipes and one to show just one. I followed what we had done in class to isolate functionality to make a recipe. |
| Describe an edge case or failure case test you wrote for your API Wrapper. |  |
| Explain how VCR aids in testing an API. | I had a really hard time with VCR testing and I'm still trying to figure it out. But in general, it records calls to the API to reuse them so that we don't have to make costly calls to the API  |
| What is the Heroku URL of your deployed application? | https://nom-app.herokuapp.com/ |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'searches#root'
+  get 'index', to: 'searches#index', as: 'searches'
+  get 'root', to: 'searches#index'
+  get ':label', to: 'searches#show', as: 'search'
+  get 'show', to: 'searches#index'
```
Comment (droberts-sea) [inline config/routes.rb:8]: It looks like you went a little overboard with routing here - you should only need 2 or 3 for this project.

File: app/views/searches/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<header>
+  <h1> <%= link_to "NOM!", root_path %></h1>
+  <%= form_tag(searches_path, method: "get", class: "header-form") do %>
+  <%= text_field_tag(:search_term, params[:search_term], class:"field") %>
+  <%= submit_tag("NOM!", class: "nom") %>
+  <% end %>
+</header>
+
+<body>
+
```
Comment (droberts-sea) [inline app/views/searches/index.html.erb:10]: A webpage should only ever have one `<body>` tag. You've got one here and one in `application.html.erb`.

File: app/views/searches/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+</header>
+
+<body>
+  <div class="show">
+
+
+  <h2> !!!! <%= @recipe.label  %> !!!! </h2>
+  <div class="info row small-up-1 large-up-2">
+    <div class="column label-image">
+      <div class = "pic">
```
Comment (droberts-sea) [inline app/views/searches/show.html.erb:16]: This is a lot of `<div>`s - this HTML might be easier to read if you were to sprinkle in some `<article>` or `<aside>`.

File: lib/muncher_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  def self.search(search_term, from = 0)
+    url = BASE_URL + "search?q=" + search_term + "&app_id=#{ID}" + "&app_key=#{KEY}" + "&from=#{from}"
+    puts "about to send request for recipes"
+    data = HTTParty.get(url)
+    puts "got response"
+    puts " status: #{data.code}: #{data.message}"
+    puts "parsed_response is #{data.parsed_response}"
+    # unless data["ok"]
+    #   raise ApiError.new("Call to list recipes failed")
+    # end
```
Comment (droberts-sea) [inline lib/muncher_api_wrapper.rb:21]: I'm not sure why you took this error handling out. If the API returns an error (for example, a new teammate is getting started and doesn't have API keys set up correctly), your app has no way to detect or report what happened.

File: lib/muncher_api_wrapper.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      end
+    end
+    return recipes_list
+  end
+
+  def self.show(uri)
+    encoded_uri = URI.encode(uri)
+    url = BASE_URL + "search?r=#{encoded_uri}" #+ uri + "&app_id=#{ID}" + "&app_key=#{KEY}"
+    data = HTTParty.get(url)
+    return create_recipe(data[0])
```
Comment (droberts-sea) [inline lib/muncher_api_wrapper.rb:35]: You should be checking the response here to make sure that the API call was successful and that a recipe was actually returned. Otherwise this will result in an error.

File: test/lib/muncher_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  it "returns an array of recipes" do
+    VCR.use_cassette("searches") do
+      result = MuncherApiWrapper.search("term")
+      result.must_be_kind_of Array
+      result.length.must_be :>, 0
+      result.each do |recipe|
+        recipe.must_be_kind_of Recipe
+      end
+    end
```
Comment (droberts-sea) [inline test/lib/muncher_api_wrapper_test.rb:18]: This is a good success case, but there are several other questions these tests aren't asking:

1. What happens if no search term is the empty string?
1. What if the search returns no results?
1. Does the optional `from` parameter change what results you get back?
1. What if the `from` parameter is invalid (negative number, etc)?
1. What if the `from` parameter takes you past the end of the available search results?

Comment (droberts-sea) [reply test/lib/muncher_api_wrapper_test.rb:18]: I would want to see some tests around the method to show an individual recipe as well. At least check it with a valid and invalid URI.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yed |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | mostly - see inline |
| Semantic HTML | some - many places use `<div>` where a more semantic tag would be appropriate |
| Errors are reported to the user | no - If a search produces no results, or especially if a recipe URI doesn't turn anything up, it would be good to report what happened to the user. |
| API Wrapper to handle the API requests | yes |
| Controller testing | no |
| Lib testing | some - these could stand to be expanded substantially |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes (no link) |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** |  |
| Link to deployed app on Heroku | yes |
|  **Overall** | This is a good start. You've got an attractive and functional website that consumes the Edamam API, and it seems like most of the core learning goals were met.<br><br>I am a little concerned about your lack of test coverage and error handling. These two skills are closely linked - they both require you to be able to take a step back and think about your method as a unit: what does success or failure look like, what combinations of inputs might cause it to fail, and what should I do when it does fail. Writing code that's stable and resilient is not easy, but it's what separates a good application from a great one. Make sure you spend some time thinking about error handling and testing with VideoStore API, and let us know if you need more support on this. |
